File: lab9KononKI305/phone.py
# phone.py
from call_history import CallHistory
from messaging import Messaging

class Phone:
    def __init__(self, brand="Unknown", model="Unknown"):
        self.brand = brand
        self.model = model
        self.battery_level = 50
        self.is_turned_on = False
        self.call_history = CallHistory()
        self.messaging = Messaging()

    def display_full_info(self):
        print("Phone Information:")
        print("Brand:", self.brand)
        print("Model:", self.model)
        print("Battery Level:", self.battery_level, "%")
        print("Power On:", "Yes" if self.is_turned_on else "No")

    def input_phone_data(self):
        self.brand = input("Enter brand: ")
        self.model = input("Enter model: ")

    def add_battery_charge(self, charge):
        if self.is_turned_on:
            self.battery_level = min(100, self.battery_level + charge)
            print("Battery charged. Current battery level:", self.battery_level, "%")
        else:
            print("Cannot charge the battery. Phone is turned off.")

    def sub_battery_charge(self, charge):
        if self.is_turned_on:
            self.battery_level = max(0, self.battery_level - charge)
            print("Battery uncharged. Current battery level:", self.battery_level, "%")
        else:
            print("Cannot uncharge the battery. Phone is turned off.")

    def turn_on(self):
        print(self.brand, self.model, "is powering on.")
        self.is_turned_on = True

    def power_off(self):
        print(self.brand, self.model, "is powering off.")
        self.is_turned_on = False

    def is_turned_on(self):
        return self.is_turned_on

    def make_call(self, phone_number):
        if self.is_turned_on:
            print("Calling", phone_number, "...")
            self.call_history.add_call(phone_number)
            self.decrease_battery_level()
        else:
            print("Phone is powered off. Cannot make a call.")

    def receive_call(self, phone_number):
        if self.is_turned_on:
            print("Incoming call from", phone_number)
            self.call_history.add_call(phone_number)
            self.decrease_battery_level()
        else:
            print("Phone is powered off. Cannot receive a call.")

    def send_message(self, phone_number, message):
        if self.is_turned_on:
            print("Sending message to", phone_number + ":", message)
            self.messaging.add_message(phone_number, message)
            self.decrease_battery_level()
        else:
            print("Phone is powered off. Cannot send a message.")

    def check_battery_level(self):
        return self.battery_level

    def decrease_battery_level(self):
        self.battery_level = max(0, self.battery_level - 5)
        print("Battery level:", self.battery_level, "%")

    def display_call_history(self):
        self.call_history.display_call_history()

    def display_messages(self):
        self.messaging.display_messages()

    def make_call_user_input(self):
        if self.is_turned_on:
            phone_number = input("Enter phone number to call: ")
            print("Calling", phone_number, "...")
            self.call_history.add_call(phone_number)
            self.decrease_battery_level()
        else:
            print("Phone is powered off. Cannot make a call.")

    def send_message_user_input(self):
        if self.is_turned_on:
            phone_number = input("Enter phone number to send a message: ")
            message = input("Enter message: ")
            print("Sending message to", phone_number + ":", message)
            self.messaging.add_message(phone_number, message)
            self.decrease_battery_level()
        else:
            print("Phone is powered off. Cannot send a message.")

    def supports_internet(self):
        raise NotImplementedError("Subclasses must implement this method")

File: lab2KononKI305/src/Phone.java
// Phone.java
import java.util.Scanner;
public class Phone {
    private String brand;
    private String model;
    private int batteryLevel;
    private boolean powerOn;
    private CallHistory callHistory;
    private Messaging messaging;
    //
    // Default constructor
    public Phone() {
        brand = "Unknown";
        model = "Unknown";
        batteryLevel = 50;
        powerOn = false;
        callHistory = new CallHistory();
        messaging = new Messaging();
    }

    // Constructor with parameters
    public Phone(String brand, String model) {
        this.brand = brand;
        this.model = model;
        batteryLevel = 100;
        powerOn = false;
        callHistory = new CallHistory();
        messaging = new Messaging();
    }

    // Display full information about the phone
    public void displayFullInfo() {
        System.out.println("Phone Information:");
        System.out.println("Brand: " + brand);
        System.out.println("Model: " + model);
        System.out.println("Battery Level: " + batteryLevel + "%");
        System.out.println("Power On: " + (powerOn ? "Yes" : "No"));
    }
    public void inputPhoneData() {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter brand: ");
        this.brand = scanner.nextLine();
        System.out.print("Enter model: ");
        this.model = scanner.nextLine();
    }

    public void addBatteryCharge(int charge) {
        if (powerOn) {
            batteryLevel = Math.min(100, batteryLevel + charge);
            System.out.println("Battery charged. Current battery level: " + batteryLevel + "%");
        } else {
            System.out.println("Cannot charge the battery. Phone is turned off.");
        }
    }
    public void subBatteryCharge(int charge) {
        if (powerOn) {
            batteryLevel = Math.min(100, batteryLevel - charge);
            System.out.println("Battery uncharged. Current battery level: " + batteryLevel + "%");
        } else {
            System.out.println("Cannot charge the battery. Phone is turned off.");
        }
    }
    // Power on the phone
    public void powerOn() {
        System.out.println(brand + " " + model + " is powering on.");
        powerOn = true;
    }

    // Power off the phone
    public void powerOff() {
        System.out.println(brand + " " + model + " is powering off.");
        powerOn = false;
    }

    // Check if the phone is powered on
    public boolean isPowerOn() {
        return powerOn;
    }

    // Make a call
    public void makeCall(String phoneNumber) {
        if (powerOn) {
            System.out.println("Calling " + phoneNumber + "...");
            callHistory.addCall(phoneNumber);
            decreaseBatteryLevel();
        } else {
            System.out.println("Phone is powered off. Cannot make a call.");
        }
    }

    // Receive a call
    public void receiveCall(String phoneNumber) {
        if (powerOn) {
            System.out.println("Incoming call from " + phoneNumber);
            callHistory.addCall(phoneNumber);
            decreaseBatteryLevel();
        } else {
            System.out.println("Phone is powered off. Cannot receive a call.");
        }
    }

    // Send a message
    public void sendMessage(String phoneNumber, String message) {
        if (powerOn) {
            System.out.println("Sending message to " + phoneNumber + ": " + message);
            messaging.addMessage(phoneNumber, message);
            decreaseBatteryLevel();
        } else {
            System.out.println("Phone is powered off. Cannot send a message.");
        }
    }

    // Check battery level
    public int checkBatteryLevel() {
        return batteryLevel;
    }

    // Decrease battery level after a certain action
    private void decreaseBatteryLevel() {
        batteryLevel -= 5;
        System.out.println("Battery level: " + batteryLevel + "%");
    }


    // Display call history
    public void displayCallHistory() {
        callHistory.displayCallHistory();
    }

    // Display messages
    public void displayMessages() {
        messaging.displayMessages();
    }
    public void makeCall() {
        if (powerOn) {
            Scanner scanner = new Scanner(System.in);
            System.out.print("Enter phone number to call: ");
            String phoneNumber = scanner.nextLine();
            System.out.println("Calling " + phoneNumber + "...");
            callHistory.addCall(phoneNumber);
            decreaseBatteryLevel();
        } else {
            System.out.println("Phone is powered off. Cannot make a call.");
        }
    }

    // Send a message by taking phone number and message from user input
    public void sendMessage() {
        if (powerOn) {
            Scanner scanner = new Scanner(System.in);
            System.out.print("Enter phone number to send a message: ");
            String phoneNumber = scanner.nextLine();
            System.out.print("Enter message: ");
            String message = scanner.nextLine();
            System.out.println("Sending message to " + phoneNumber + ": " + message);
            messaging.addMessage(phoneNumber, message);
            decreaseBatteryLevel();
        } else {
            System.out.println("Phone is powered off. Cannot send a message.");
        }
    }
}
